lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+        if (!$fragment){
+            return
+        }
+        $details = $fragment.Split("=")
+        $algorithm = $details[0]
+        $hash = $details[1]
+        if($algorithm -in @("SHA1", "SHA256", "SHA384", "SHA512", "MACTripleDES", "MD5", "RIPEMD160")){
+            Test-FileIntegrity -File $OutFile -Algorithm $algorithm -ExpectedHash $hash
+        } else {
+            Throw "Hash algorithm $algorithm not recognized."
+        }
+    }
+}
+
+
+# Returns the full path of the package after it is downloaded using
+# the URL parameter (a checksum may optionally be specified). The
+# package is cached on the disk until the installation successfully finishes.
+# If the hook fails, on the second run this function will return the cached
+# package path if checksum is given and it matches.
+function Get-PackagePath {
+    Param(
+        [Parameter(Mandatory=$True)]
+        [string]$URL,
+        [string]$Checksum="",
+        [string]$HashingAlgorithm=""
+    )
+
+    $packagePath = Join-Path $env:TEMP $URL.Split('/')[-1]
+    if (Test-Path $packagePath) {
+        if ($Checksum -and $HashingAlgorithm) {
+            if (Test-FileIntegrity -File $packagePath -Algorithm $HashingAlgorithm -ExpectedHash $Checksum) {
+                return $packagePath
+            }
+        }
+        Remove-Item -Recurse -Force -Path $packagePath
+    }
+
+    if ($Checksum -and $HashingAlgorithm) {
+        Start-DownloadFile -Uri "$URL#$HashingAlgorithm=$Checksum" -OutFile $packagePath
+    } else {
+        Start-DownloadFile -Uri $URL -OutFile $packagePath -SkipIntegrityCheck
+    }
+
+    return $packagePath
+}
+
+
+# Installs a package after it is downloaded from the Internet and checked for
+# integrity with SHA1 checksum. Accepts as parameters: an URL, an optional
+# 'Checksum' with its 'HashingAlgorithm' and 'ArgumentList' which can be passed
+# if the installer requires unattended installation.
+# Supported packages formats are: '.exe' and '.msi'
+function Install-Package {
+    Param(
+        [string]$URL,
+        [string]$Checksum="",
+        [string]$HashingAlgorithm="",
+        [array]$ArgumentList
+    )
+
+    Write-JujuLog "Installing package '$URL'"
+
+    $packageFormat = $URL.Split('.')[-1]
+    $acceptedFormats = @('msi', 'exe')
+    if ($packageFormat -notin $acceptedFormats) {
+        Throw ("Cannot install the package found at this URL: $URL " +
+               "Unsupported installer format.")
+    }
+
+    $installerPath = Get-PackagePath $URL $Checksum $HashingAlgorithm
+    $stat = Start-Process -FilePath $installerPath -ArgumentList $ArgumentList `
+                          -PassThru -Wait
+    if ($stat.ExitCode -ne 0) {
+        throw "Package '$URL' failed to install."
+    }
+    Remove-Item $installerPath
+
+    Write-JujuLog "Finished installing package."
+}
+
+
+function Start-GitClonePull {
+    Param(
+        [string]$Path,
+        [string]$URL,
+        [string]$Branch="master"
+    )
+
```

Comment:
So there are a few points here:

First, I would define any interesting variables inside the function, such as the actual joined path and the path to the ```.git``` dir.

```powershell
$projectDir = $Path
$gitPath = Join-Path $projectDir ".git"
```